src/core/utils/components/VirtualScroll.tsx

```diff
@@ -0,0 +1,232 @@
+ * Currently it works only when items direct parent is a scrollable container (body scroll is not supported yet).
+ * It supports only static (same) height rows virtualization. Expect some issues, if list consists of different height elements.
+ * @example
+ * const itemRenderer = React.useCallback(() => (
+ *  <div key={index}>
+ *    This is my item #{index}
+ *  </div>
+ * ), [])
+ * <VirtualScroll
+ *  itemsLength={1000}
+ *  itemRenderer={itemRenderer}
+ * />
+ */
+export const VirtualScroll = ({
+  itemsLength: childrenLength,
+  itemRenderer,
+  renderAdditional = 20,
+}: VirtualScrollProps) => {
+  const [startNode, setStartNode] = React.useState(0);
+  const [visibleNodeCount, setVisibleNodeCount] = React.useState(0);
+  const scrollContainer = React.useRef<HTMLElement>();
+  const childrenParentRef = React.useRef<HTMLDivElement>(null);
+  const childHeight = React.useRef(0);
+  const [translateY, setTranslateY] = React.useState(0);
+  const onScrollRef = React.useRef<(e: Event) => void>();
+
+  // Find scrollable parent
+  // Needed only on init
+  React.useLayoutEffect(() => {
+    const scrollableParent = getScrollableParent(childrenParentRef.current);
+    scrollContainer.current = scrollableParent;
+  }, []);
+
+  const visibleChildren = React.useMemo(() => {
+    const arr = [];
+    const endIndex = Math.min(
+      childrenLength,
+      startNode + visibleNodeCount + renderAdditional,
+    );
+    for (let i = startNode; i < endIndex; i++) {
+      arr.push(itemRenderer(i));
+    }
+    return arr;
+  }, [
+    childrenLength,
+    itemRenderer,
+    renderAdditional,
+    startNode,
+    visibleNodeCount,
+  ]);
+
+  // Get child height when children available
+  React.useLayoutEffect(() => {
+    if (!childrenParentRef.current || !visibleChildren.length) {
+      return;
+    }
+    let heightSum = 0;
+    for (let i = startNode; i < visibleChildren.length + startNode; i++) {
+      const htmlElement = childrenParentRef.current.children.item(
+        Math.max(0, i - startNode),
+      ) as HTMLElement;
+      const elementHeight = getElementHeight(htmlElement);
+      heightSum += Number(elementHeight.toFixed(2));
+    }
+    const elementHeight = heightSum / visibleChildren.length;
+    childHeight.current = Math.ceil(elementHeight);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [visibleChildren.length]);
+
+  const onScroll = React.useCallback(
+    (e: Event) => {
+      const target =
+        scrollContainer.current !== document.body
+          ? (e.target as HTMLElement)
+          : (document.scrollingElement as HTMLElement);
+      if (!target) {
+        return;
+      }
+      const start = getCountOfNodesInHeight(
+        childHeight.current,
+        target.scrollTop,
+      );
+      setStartNode(start);
+      setVisibleNodeCount(
+        getVisibleNodeCount(childHeight.current, start, childrenLength, target),
+      );
+      setTranslateY(getTranslateValue(childHeight.current, start));
```

Comment:
I am thinking, what if you changed translate value directly in the ref? It should be a bit faster as it avoids rerenders from setState. But maybe not a big performance gain as it would only work when startNode and visibleCount is the same.

Reply:
using ref now, seems like still working